account_analytic_default_account/models/account_analytic_default_account.py

```diff
@@ -0,0 +1,130 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+
+from odoo import api, fields, models
+
+
+class AccountAnalyticDefaultAccount(models.Model):
+    _inherit = "account.analytic.default"
+
+    account_id = fields.Many2one(
+        'account.account', string='Account',
+        ondelete='cascade',
+        help="Select an account which will use analytic account specified in "
+             "analytic default (e.g. create new customer invoice or "
+             "Sales order if we select this product, "
+             "it will automatically take this as an analytic account)"
+    )
+
+    @api.model
+    def account_get(self, product_id=None, partner_id=None, user_id=None,
```

Comment:
I'd split this in more methods and simplify it a bit, for instance:
```python
    account_get_domain_keys = (
        'product_id', 'partner_id', 'user_id',
        'date', 'company_id', 'account_id',
    )

    def _account_get_domain(self, **kw):
        """Build account.analytic.default domain.

        :param kw: filter keys.
            Available filter keys are defined into `account_get_domain_keys`.
        :return: a search domain with all required keys' leaves.
            Each key will be searched for both False and specified value.
            For instance: [
                ('product_id'), '=', False), '|', [('product_id'), '=', 100)
            ]
        """
        domain = []
        account_get_domain_keys = self.account_get_domain_keys[:]
        date_key = account_get_domain_keys.pop('date', None)
        for key in account_get_domain_keys:
            if kw.get(key):
                domain += ['|', (key, '=', kw.get(key))]
            domain += [(key, '=', False)]
        if date_key and kw.get(date_key):
            domain += ['|', ('date_start', '<=', kw.get(date_key)),
                       ('date_start', '=', False)]
            domain += ['|', ('date_stop', '>=', kw.get(date_key)),
                       ('date_stop', '=', False)]
        return domain

    def account_get(self, **kw):
        domain = self._account_get_domain(**kw)
        best_index = -1
        res = self.env['account.analytic.default']
        for rec in self.search(domain):
            index = 0
            for k in kw.keys():
                if rec[k]:
                    index += 1
            if index > best_index:
                res = rec
                best_index = index
        return res
```

Reply:
I can't change the method signature so drastically as it already exists in the Odoo addons. I can't guarantee that the calls are made with kwargs everywhere. That's the main reason I copied the method and updated it for my needs.

Reply:
fine, you are right. Keep the signature as it is... but respect it in your calls if possible :)
Then you can still address the rest I guess ;)